pennylane/optimize/gradient_descent.py

```diff
@@ -79,14 +96,23 @@ def compute_grad(objective_fn, x, grad_fn=None):
                 If ``None``, the gradient function is computed automatically.
 
         Returns:
-            array: NumPy array containing the gradient :math:`\nabla f(x^{(t)})`
+            tuple: tuple with NumPy array containing the gradient :math:`\nabla f(x^{(t)})` and the
+                objective function output
         """
         if grad_fn is not None:
-            g = grad_fn(x)  # just call the supplied grad function
+            g = grad_fn  # just call the supplied grad function
+
+            try:
+                forward = g.forward
+            # if grad_fn is not an autograd gradient function, calculate the
+            # objective function output separately
+            except AttributeError:
+                forward = objective_fn(x)
```

Comment:
I'm wondering if this makes sense. if `grad_fn` is not an autograd gradient function, perhaps we just return `None` for the forward pass value.

With autograd, there is no additional cost to computing the forward pass -- it is there automatically. However, if not using autograd, we could just not support this behaviour to avoid an unnecessary forward pass happening (that the user might not want/need).

Reply:
This would also simply the logic a lot. We could simply do
```python
def compute_grad(objective_fn, x, grad_fn=None):
    g = get_gradient(objective_fn) if grad_fn is None else grad_fn
    grad = g(x)
    forward = getattr(g, "forward", None)
    return grad, forward
```

Reply:
I like this! It's makes more sense and avoids having to compute the output if there's no forward pass. It's then left for the user to calculate the cost separately in those cases.